Introduce SQLite application (#42)
unikraft/catalog-core: sqlite/README.md

# SQLite on Unikraft

Build and run SQLite on Unikraft.
Follow the instructions below to set up, configure, build and run SQLite.
Make sure you installed the [requirements](../README.md#requirements).

## Quick Setup (aka TLDR)

For a quick setup, run the commands below.
Note that you still need to install the [requirements](../README.md#requirements).
Before everything, make sure you run the [top-level `setup.sh` script](../setup.sh).

To build and run the application for `x86_64`, use the commands below:

```console
./setup.sh
make distclean
wget -O /tmp/defconfig https://raw.githubusercontent.com/unikraft/catalog-core/refs/heads/scripts/sqlite/scripts/defconfig/qemu.x86_64
UK_DEFCONFIG=/tmp/defconfig make defconfig
make -j $(nproc)
test -f initrd.cpio || ./workdir/unikraft/support/scripts/mkcpio initrd.cpio ./rootfs/
qemu-system-x86_64 \
	-nographic \
	-kernel workdir/build/sqlite_qemu-x86_64 \
	-append "sqlite_qemu-x86_64 vfs.fstab=[ \"initrd0:/:extract::ramfs=1:\" ] -- /chinook.db 'SELECT * FROM Album LIMIT 10'" \

> 	-initrd ./initrd.cpio

Reviewer: Do not use Tabs. Use space / blank.

```

This will configure, build and run SQLite on Unikraft.
You can see how to test it in the ["Test" section](#test).
To close the virtual machine, see the instructions in the ["Close QEMU" section](#close-qemu).

To do the same for `AArch64`, run the commands below:

```console
./setup.sh
make distclean
wget -O /tmp/defconfig https://raw.githubusercontent.com/unikraft/catalog-core/refs/heads/scripts/sqlite/scripts/defconfig/qemu.arm64
UK_DEFCONFIG=/tmp/defconfig make defconfig
make -j $(nproc)
test -f initrd.cpio || ./workdir/unikraft/support/scripts/mkcpio initrd.cpio ./rootfs/
qemu-system-aarch64 \
	-nographic \
	-machine virt \
	-cpu max \
	-kernel workdir/build/sqlite_qemu-arm64 \
	-append "sqlite_qemu-arm64 vfs.fstab=[ \"initrd0:/:extract::ramfs=1:\" ] -- /chinook.db 'SELECT * FROM Album LIMIT 10'" \

> 	-initrd ./initrd.cpio

Reviewer: Do not use Tabs. Use space / blank.

And so on.

```

Similar to the `x86_64` build, this will configure, build and run SQLite on Unikraft.

> [!TIP]

> > To make the SQLite session interactive, as opposed to running a specific query, simply remove the `SELECT * FROM Album LIMIT 10` part from the `-append` argument. This will allow you to interact with the SQLite database directly using the shell. The argument would be:

Reviewer: ```suggestion
> To make the SQLite session interactive, as opposed to running a specific query, remove the `SELECT * FROM Album LIMIT 10` part from the `-append` argument.
> The only argument will the the `/chinook.db` path to the database file.
> This will allow you to interact with the SQLite database directly using the shell.
> The argument would be:
```

> ```bash
> -append "sqlite_qemu-arm64 vfs.fstab=[ \"initrd0:/:extract::ramfs=1:\" ] -- /chinook.db" \
> ```
>  This removes the fixed query, allowing for full interactive database access.

Information about every step and about other types of builds is detailed below.

## Set Up

Set up the required repositories.
For this, you have two options:

1. Use the `setup.sh` script:

```console
./setup.sh
```

It will create symbolic links to the required repositories in `../repos/`.
Be sure to run the [top-level `setup.sh` script](../setup.sh).

> If can use a custom variant of repositories (e.g. apply your own patch, make modifications); update it accordingly in the `../repos/` directory.

Reviewer: Indent contents to align with the `Use the setup.sh script` line. Basically, add 3 blanks / spaces.

1. Have your custom setup of repositories in the `workdir/` directory.

> Clone, update and customize repositories to your own needs.

Reviewer: ```suggestion
   Clone, update and customize repositories to your own needs.
```

Indent to align correctly for a list.

```console
make properclean
make CC=clang -j $(nproc)
```

Note that Clang >= 14 is required to build Unikraft.

To build with another GCC version, use the commands below:

```console
make properclean
make CC=gcc-<version> -j $(nproc)
```

where `<version>` is the GCC version, such as `11`, `12`.

Note that GCC >= 8 is required to build Unikraft.

### Build the Filesystem

The filesystem is to be packed into `initrd.cpio`, an initial ramdisk CPIO file.
Use the command below for that:

```console
rm -f initrd.cpio
./workdir/unikraft/support/scripts/mkcpio initrd.cpio ./rootfs/
```

## Run

Run the resulting image using the corresponding platform tool.
Firecracker requires KVM support.
Xen requires a system with Xen installed.

A successful run will show a message such as the one below:

```text
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
                 Helene 0.18.0~12072b5f
-- warning: cannot find home directory; cannot read ~/.sqliterc
SQLite version 3.40.1 2022-12-28 14:03:47
Enter ".help" for usage hints.

> sqlite>

Reviewer: Show the actual output of running SQLite with the `SELECT query`.

  -kernel workdir/build/sqlite_qemu-arm64 \
  -append "sqlite_qemu-arm64 vfs.fstab=[ \"initrd0:/:extract::ramfs=1:\" ] -- /chinook.db 'SELECT * FROM Album LIMIT 10'" \
  -initrd ./initrd.cpio
```

Similar to the `x86_64` build, this will configure, build and run SQLite on Unikraft.

> [!TIP]
> To make the SQLite session interactive, as opposed to running a specific query, remove the `SELECT * FROM Album LIMIT 10` part from the `-append` argument.
> The only argument will be the `/chinook.db` path to the database file.
> This will allow you to interact with the SQLite database directly using the shell.
> The argument would be:
> ```bash
> -append "sqlite_qemu-arm64 vfs.fstab=[ \"initrd0:/:extract::ramfs=1:\" ] -- /chinook.db" \
> ```
> This removes the fixed query, allowing for full interactive database access.

Information about every step and about other types of builds is detailed below.

## Set Up

Set up the required repositories.
For this, you have two options:

1. Use the `setup.sh` script:

   ```console
   ./setup.sh
   ```

   It will create symbolic links to the required repositories in `../repos/`.
   Be sure to run the [top-level `setup.sh` script](../setup.sh).

>    If you want you can use a custom variant of repositories (e.g. apply your own patch, make modifications); update it accordingly in the `../repos/` directory.

Reviewer: ```suggestion
   If you want you can use a custom variant of repositories (e.g. apply your own patch, make modifications), update it accordingly in the `../repos/` directory.
```

For consistency with the other repos.

```

where `<version>` is the GCC version, such as `11`, `12`.

Note that GCC >= 8 is required to build Unikraft.

### Build the Filesystem

The filesystem is to be packed into `initrd.cpio`, an initial ramdisk CPIO file.
Use the command below for that:

```console
rm -f initrd.cpio
./workdir/unikraft/support/scripts/mkcpio initrd.cpio ./rootfs/
```

## Run

Run the resulting image using the corresponding platform tool.
Firecracker requires KVM support.
Xen requires a system with Xen installed.

A successful run will show a message such as the one below:

```text
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
                    Pan 0.19.1~d513663f
-- warning: cannot find home directory; cannot read ~/.sqliterc
1|For Those About To Rock We Salute You|1
2|Balls to the Wall|2
3|Restless and Wild|2
4|Let There Be Rock|1
5|Big Ones|3
6|Jagged Little Pill|4
7|Facelift|5
8|Warner 25 Anos|6
9|Plays Metallica By Four Cellos|7
10|Audioslave|8
```

This means that SQLite runs on Unikraft waiting for further instructions.

> 

Reviewer: Mention that the line: "-- warning: cannot find home directory; cannot read ~/.sqliterc" is expected because of a missing configuration file for SQLite - an optional file.

Author: Added a sensible `.sqliterc` in the file system. Therefore this is now solved.

### Run on QEMU/x86_64

To run the Unikraft image:

```console
qemu-system-x86_64 \
  -nographic \
  -kernel workdir/build/sqlite_qemu-x86_64 \
  -append "sqlite_qemu-x86_64 vfs.fstab=[ \"initrd0:/:extract::ramfs=1:\" ] -- /chinook.db 'SELECT * FROM Album LIMIT 10'" \

>   -initrd ./initrd.cpio

Reviewer: Indent lines starting from the 2nd one with 4 blanks. Be consistent with the other apps.

sudo xl create -c xen.x86_64.cfg
```

You need use `sudo` or the `root` account to run Xen.

### Run on Xen/ARM64

To run the Unikraft image:

```console
sudo xl create -c xen.arm64.cfg
```

You need use `sudo` or the `root` account to run Xen.

## Test

### Close QEMU

To close the QEMU virtual machine, use the `Ctrl+a x` keyboard shortcut;
that is press the `Ctrl` and `a` keys at the same time and then, separately, press the `x` key.

### Close Firecracker

To close the Firecracker virtual machine, open another console and use the command:

```console
sudo pkill -f firecracker
```

### Close Xen

To close the Xen virtual machine, open another console and use the command:

```console
sudo xl destroy sqlite
```

## Clean Up

Doing a new configuration, or a new build, may require cleaning up the configuration and build artifacts.

In order to remove the build artifacts, use:

```console
make clean
```

In order to remove fetched files also, that is the removal of the `workdir/build/` directory, use:

```console
make properclean
```

In order to remove the generated `.config` file as well, use:

```console
make distclean
```

## Customize

### Customize the Filesystem Contents

> The main way to customize SQLite is to update its configuration in the directory of the filesystem:

Reviewer: These instructions don't tell me anything. What can I do to update the configuration? What is the `script.sql` file? It it used anywhere?

```console
rootfs/
|-- chinook.db
`-- script.sql
```

After updating the filesystem contents, you need to [rebuild the filesystem](#build-the-filesystem) and the [run](#run) the unikernel.
No rebuild of the unikernel is necessary.

### Use a Different Filesystem Type for QEMU

You can use [`9pfs`](https://github.com/unikraft/unikraft/tree/staging/lib/9pfs) as an alternate filesystem to initrd.
Note that 9pfs does not work with Firecracker.
And it requires re-building Xen to add 9pfs support.
Below find instructions on running SQLite on QEMU with 9pfs support.

You need to use these contents for the [`Config.uk`](Config.uk) configuration file:

```text
config SQLITE
bool "Configure SQLite application with 9pfs as rootfs"
default y
  select LIBSQLITE
  select LIBSQLITE_MAIN_FUNCTION
  select LIBVFSCORE
  select LIBVFSCORE_AUTOMOUNT_UP
  select LIBUK9P

>   select LIB9PFS

Reviewer: Indent lines starting from the 2nd one with 8 blanks. Be consistent with the other apps.

You can use [`9pfs`](https://github.com/unikraft/unikraft/tree/staging/lib/9pfs) as an alternate filesystem to initrd.
Note that 9pfs does not work with Firecracker.
And it requires re-building Xen to add 9pfs support.
Below find instructions on running SQLite on QEMU with 9pfs support.

You need to use these contents for the [`Config.uk`](Config.uk) configuration file:

```text
config SQLITE
bool "Configure SQLite application with 9pfs as rootfs"
default y
        select LIBSQLITE
        select LIBSQLITE_MAIN_FUNCTION
        select LIBVFSCORE
        select LIBVFSCORE_AUTOMOUNT_UP
        select LIBUK9P
        select LIB9PFS
```

This means you replace these two lines in [`Config.uk`](Config.uk):

```text
        select LIBRAMFS
        select LIBUKCPIO
```

with these lines:

```text
        select LIBUK9P
        select LIB9PFS
```

Now go through the [configure](#configure) and [build](#build) steps.

Finally, use the commands below to run SQLite on QEMU/x86_64 with 9pfs support:

```console
# Create a copy of the filesystem to be mounted as 9pfs.
rm -fr 9pfs-rootfs
cp -r rootfs 9pfs-rootfs
sudo qemu-system-x86_64 \
    -nographic \
    -cpu max \

>     -append "sqlite_qemu-x86_64 vfs.fstab=[ \"initrd0:/:extract::ramfs=1:\" ] env.vars=[ \"HOME=/root\" ] -- /chinook.db 'SELECT * FROM Album LIMIT 10'" \

Reviewer: ```suggestion
    -append "sqlite_qemu-x86_64 vfs.fstab=[ \"fs0:/:9pfs:::\" ] env.vars=[ \"HOME=/root\" ] -- /chinook.db 'SELECT * FROM Album LIMIT 10'" \
```
